### ISSUE TITLE
[AM][gateway] struggle to make the role mapper work with LDAP IdP (and dynamic scopes)

### ISSUE THREAD

**Issue by bcollard** [gravitee-io/issues#3530]
When using a LDAP as the IdP, I can't make the role mapper work as expected.

### Given the following config for the Domain

**Domain**

![image](https://user-images.githubusercontent.com/9626829/79123404-5f911280-7d9a-11ea-89ad-1e99645f8781.png)

**Provider**

![image](https://user-images.githubusercontent.com/9626829/79123391-5b64f500-7d9a-11ea-832c-96c198baae7c.png)

**Role**

![image](https://user-images.githubusercontent.com/9626829/79121109-d4614e00-7d94-11ea-82e4-f36d55d4cd30.png)

**LDAP provider role mapping**

![image](https://user-images.githubusercontent.com/9626829/79121090-c6abc880-7d94-11ea-8e3f-cd70b7d7dbcb.png)

**Client settings (in domain config)**

![image](https://user-images.githubusercontent.com/9626829/79122298-a0d3f300-7d97-11ea-8d95-5289d30d77a6.png)

### For the Client

**Client config**

![image](https://user-images.githubusercontent.com/9626829/79121209-0ecaeb00-7d95-11ea-855b-1707e5f45f29.png)

If the user is member of the `admins` group, then he should be affected the role "APP_ADMIN" and the scope "APP_ADMIN" should be added to the OAuth response.

## Current Behavior

The role is not attached to the User.

```
{
	"_id" : "47ad494a-7950-4426-ad49-4a7950b42608",
	"accountNonExpired" : true,
	"accountNonLocked" : true,
	"additionalInformation" : {
		"sub" : "jdoe",
		"auth_time" : NumberLong(1586780468),
		"preferred_username" : "jdoe",
		"family_name" : "John Doe",
		"email" : "john.doe@mycompany.com"
	},
	"client" : "699a6a52-b5d7-497a-9a6a-52b5d7c97ad5",
	"createdAt" : ISODate("2020-04-13T12:21:08.874Z"),
	"credentialsNonExpired" : true,
	"domain" : "oidc-ldap",
	"enabled" : true,
	"externalId" : "jdoe",
	"internal" : false,
	"loggedAt" : ISODate("2020-04-13T12:21:08.874Z"),
	"loginsCount" : NumberLong(1),
	"preRegistration" : false,
	"registrationCompleted" : false,
	"roles" : [ ],
	"source" : "db9c78bb-e79b-4998-9c78-bbe79bc998fc",
	"updatedAt" : ISODate("2020-04-13T12:21:17.900Z"),
	"username" : "jdoe"
}
```

The dynamic scope is not added to the response payload.

## Steps to Reproduce (for bugs)
1. OPTIONAL: git pull that openldap docker repo: https://github.com/bcollard/simple-openldap-docker (so that you see how the image run in the next step and you can have a look to the LDIF)
2. use that modified docker-compose.yml with the LDAP service added: https://gist.github.com/bcollard/881b74433b3ea8bb85448fef67cbfb4a
3. run Apache Directory Studio and have a look to the LDAP structure (localhost:389 ; bind DN: `cn=gravitee-am,ou=service-accounts,dc=mycompany,dc=corp` bind pwd: `gravitee-am` base DN: `dc=mycompany,dc=corp`)
4. configure the AM as described above
5. visit https://oidcdebugger.com/ and configure the OIDC client like so:
![image](https://user-images.githubusercontent.com/9626829/79121974-e2b06980-7d96-11ea-8fa9-0384d081903e.png)

Response type (required): code
Response mode (required): query

6. auth with "jdoe" / pwd "john" or with "jsmith", pwd "jack"

7. run this command
`curl -X POST "http://localhost/am/oidc-ldap/oauth/token?grant_type=authorization_code&client_id=testing-client&client_secret=testing-client&redirect_uri=https://oidcdebugger.com/debug&code=<code returned by OIDC debugger>`

8. as you can see in the reponse payload, the scope is not added.
Also, the role is not even present in the User Profile (Users collection in mongo)

## Your Environment
* Version used: 2.10.11

**Comment by tcompiegne** [comment]
Hi @bcollard , did you try with the following LDAP IdP configuration : 

Group search filter : `(member={0})` ?

**Comment by bcollard** [comment]
damn it!
yes, there was my mistake! thank you!

However, could you please clarify how is attributed the role please? I thought it was linked to the expression I use in the role mapper, like `memberOf=admins`
And the fact is it's linked to the LDAP configuration for groups. So i guess it's both, correct me if I'm wrong.

**Comment by tcompiegne** [comment]
`However, could you please clarify how is attributed the role please ?`

1. Authenticate the user (LDAP Bind or Compare operation)
2. Fetch user's groups (at this step we are using the group search filter) 
  => if the user belongs to some groups add the `memberOf` attribute to the user with `cn` group name
3. Apply user mapper
4. Apply role mapper

**Comment by bcollard** [comment]
Thank you @tcompiegne 

As far as I know, the _memberOf_ attribute is optional (and operational). Its declaration should be done through the _memberOf_ overlay, when enabled.

From your explanations, I understand that you set the value of _memberOf_ to the cn value of the group. It should be [the DN](https://www.openldap.org/doc/admin24/overlays.html#Reverse%20Group%20Membership%20Maintenance) of the group instead.

And regarding the role mapper, I was expecting this kind of condition to work: `memberOf=cn=admins,ou=groups,dc=mycompany,dc=corp`

And that the group configuration in the LDAP IdP config would be for group management purposes ("Groups" tab), ie. not linked to roles mapping.

As I'm not anymore stuck by this issue, we can close it; or keep it open if you think it's worth it.

**Comment by lazki** [comment]
Thank @bcollard for your detailed issue report. Help me understand the Role/Scope Mapping.

@tcompiegne it is not easy for people not familiar with Spring Security
https://docs.spring.io/spring-security/site/docs/3.2.0.CI-SNAPSHOT/reference/html/ldap.html
to understand the Role/Scope mapping. The key was the `group-role-attribute`.

At start, I was expecting the evaluation of `memberOf` to use the DN _too_.  But AM is using the `group-role-attribute` value defined above.

Without this Issue, I will never learn how to use this feature. 

I think when you use `memberOf` you should use the full DN to evaluate membership.
Ref: [memberOf](https://ldapwiki.com/wiki/MemberOf)

This Issue should be remain open as a reminder to improve documentation.
https://docs.gravitee.io/am/2.x/am_userguide_identity_provider_mapping.html

**Comment by tcompiegne** [comment]
Hi @lazki , we do not use Spring Security for the LDAP authentication. 

I was not aware of this documentation :  https://ldapwiki.com/wiki/MemberOf

Indeed It can be better to change the `cn` by the `dn` attribute, but you can already do that in the identity provider configuration.

**Comment by lazki** [comment]
@tcompiegne you're right. I changed the `cn` to `dn` and the evaluation of `memberOf`worked as expected.

@bcollard should notice this.
